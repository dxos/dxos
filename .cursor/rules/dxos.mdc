---
description: Working in the DXOS codebase
globs: 
alwaysApply: false
---
# DXOS Rules

Key preferences:
- Write in TypeScript by default.
- Prefer functional programming patterns.
- Follow React best practices and hooks patterns.
- Use TailwindCSS for React styles.
- Unit test examples when relevant.

When writing code:
- Include type definitions.
- Add JSDoc comments for public functions.
- Keep components pure and composable.
- Use proper error handling with Effect.
- Include proper event handler types for React components.
- All comments should be in English and end with a period.

Technologies to consider:
- Use Effect-TS for schema validation.
- Automerge for state management.
- Codemirror v6 for editor integrations.

Monorepo tools:
- moon for monorepo management.
- pnpm for package management.
- vitest
- vite

Testing:
- Place test files near the module they test with "module_name.test.ts" format.
- Use vitest with `describe` and `test` apis (don't use `it`).
- Don't write tests at top-level, instead make sure they are wrapped in a `describe` block.
- Prefer importing `expect` from test context: `test('foo', ({ expect }) => ...)`