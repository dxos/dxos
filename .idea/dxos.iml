<?xml version="1.0" encoding="UTF-8"?>
<module type="WEB_MODULE" version="4">
  <component name="NewModuleRootManager">
    <content url="file://$MODULE_DIR$">
      <excludeFolder url="file://$MODULE_DIR$/tools/executors/protobuf-compiler/test/__snapshots__" />
      <excludeFolder url="file://$MODULE_DIR$/packages/common/protocols/proto" />
      <excludePattern pattern="gen" />
      <excludePattern pattern="out" />
      <excludePattern pattern="tmp" />
    </content>
    <orderEntry type="inheritedJdk" />
    <orderEntry type="sourceFolder" forTests="false" />
  </component>
</module>
