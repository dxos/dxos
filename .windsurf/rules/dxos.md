---
trigger: model_decision
description: Working in the DXOS codebase
---

# DXOS Rules

Key preferences:

- Write in TypeScript by default.
- Prefer functional programming patterns.
- Follow Effect TS best practices for function patterns.
- Follow React best practices and hooks patterns.
- Use TailwindCSS for React styles.
- Unit test examples when relevant.

When writing code:

- Include type definitions.
- Add JSDoc comments for public functions.
- Keep components pure and composable.
- Use proper error handling with Effect.
- Include proper event handler types for React components.
- All comments should be in English and end with a period.

Variable names:

- Use \_ as a prefix if the variable is not used.
- Use $ as a prefix to disambiguate with local variables.

Technologies to consider:

- Use Effect-TS for schema validation.
- Automerge for state management.
- Codemirror v6 for editor integrations.

Monorepo tools:

- moon for monorepo management.
- pnpm for package management.
- vitest
- vite

Testing:

- Place test files near the module they test with "module_name.test.ts" format.
- Use vitest with `describe` and `test` apis (don't use `it`).
- Don't write tests at top-level, instead make sure they are wrapped in a `describe` block.
- Prefer importing `expect` from test context: `test('foo', ({ expect }) => ...)`
