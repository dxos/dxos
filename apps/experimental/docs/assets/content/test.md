---
type: 'content'
path: 'section.test'
---

# TEST

Your mission is to create the first functionally and commercially competitive, 
fully decentralized alternative to the public cloud.

## Compete with the Cloud

It is not enough to protest extreme centralization; what matters is to replace it with a user-sovereign, privacy-preserving alternative. This alternative must enable applications that can compete with the cloud-based incumbents that dominate the application landscape.


## Build New Kinds of Applications

DXOS enables the development of entirely new classes of decentralized applications. It is a network operating system, infrastructure platform, programming framework, and toolchainâ€”providing a complete development solution that rivals conventional cloud platforms. 


## Protect Privacy and Data Sovereignty

Data sovereignty and privacy are enforced at the protocol level by a uniquely scalable decentralized database. Unlike conventional approaches, data is not held captive by applications. Instead, individuals and teams are able to select and change which applications have access to their data. 
