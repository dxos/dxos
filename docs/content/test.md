# Test

This is an unlisted page which may be useful for testing new docs functionality.
