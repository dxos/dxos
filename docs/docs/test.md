# Test

This is an unlisted page which may be useful for testing new docs functionality.

```tsx file=../src/demos/Test.tsx#L13-L37 showcase peers=2 controls=airplane setup=identity,space
  const tasks = useQuery(space, Task.filter());
  const [input, setInput] = useState<HTMLInputElement>();

  const handleKeyDown: KeyboardEventHandler<HTMLInputElement> = async (event) => {
    if (event.key === 'Enter' && input) {
      const task = new Task({ title: input.value });
      input.value = '';
      await space.db.add(task);
    }
  };

  return (
    <div>
      <input ref={setInput} onKeyDown={handleKeyDown} />
      {tasks.map((task) => (
        <div key={task.id}>
          <input type='checkbox' checked={!!task.completed} onChange={() => (task.completed = !task.completed)} />
          {task.title}
          <button onClick={() => space.db.remove(task)}>x</button>
        </div>
      ))}
    </div>
  );
};

```
