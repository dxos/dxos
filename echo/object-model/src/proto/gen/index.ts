import { Schema } from "@dxos/codec-protobuf";
import * as dxos_echo_object from "./dxos/echo/object";
export interface TYPES {
    "dxos.echo.object.Array": dxos_echo_object.Array;
    "dxos.echo.object.KeyValue": dxos_echo_object.KeyValue;
    "dxos.echo.object.Object": dxos_echo_object.Object;
    "dxos.echo.object.ObjectMutation": dxos_echo_object.ObjectMutation;
    "dxos.echo.object.ObjectMutation.Operation": dxos_echo_object.ObjectMutation.Operation;
    "dxos.echo.object.ObjectMutationSet": dxos_echo_object.ObjectMutationSet;
    "dxos.echo.object.ObjectSnapshot": dxos_echo_object.ObjectSnapshot;
    "dxos.echo.object.Predicate": dxos_echo_object.Predicate;
    "dxos.echo.object.Predicate.Operation": dxos_echo_object.Predicate.Operation;
    "dxos.echo.object.Query": dxos_echo_object.Query;
    "dxos.echo.object.Value": dxos_echo_object.Value;
}
export const schemaJson = JSON.parse("{\"nested\":{\"dxos\":{\"nested\":{\"echo\":{\"nested\":{\"object\":{\"nested\":{\"Array\":{\"fields\":{\"values\":{\"rule\":\"repeated\",\"type\":\"Value\",\"id\":1}}},\"KeyValue\":{\"fields\":{\"key\":{\"type\":\"string\",\"id\":1},\"value\":{\"type\":\"Value\",\"id\":2}}},\"Object\":{\"fields\":{\"properties\":{\"rule\":\"repeated\",\"type\":\"KeyValue\",\"id\":1}}},\"ObjectMutation\":{\"fields\":{\"operation\":{\"type\":\"Operation\",\"id\":1},\"key\":{\"type\":\"string\",\"id\":2},\"value\":{\"type\":\"Value\",\"id\":3}},\"nested\":{\"Operation\":{\"values\":{\"SET\":0,\"DELETE\":1,\"ARRAY_PUSH\":2,\"SET_ADD\":4,\"SET_DELETE\":5}}}},\"ObjectMutationSet\":{\"fields\":{\"mutations\":{\"rule\":\"repeated\",\"type\":\"ObjectMutation\",\"id\":1}}},\"ObjectSnapshot\":{\"fields\":{\"root\":{\"type\":\"Value\",\"id\":1}}},\"Predicate\":{\"fields\":{\"op\":{\"type\":\"Operation\",\"id\":1},\"key\":{\"type\":\"string\",\"id\":2},\"value\":{\"type\":\"Value\",\"id\":3},\"predicates\":{\"rule\":\"repeated\",\"type\":\"Predicate\",\"id\":4}},\"nested\":{\"Operation\":{\"values\":{\"OR\":0,\"AND\":1,\"NOT\":2,\"IN\":10,\"EQUALS\":11,\"GTE\":12,\"GT\":13,\"LTE\":14,\"LT\":15,\"PREFIX_MATCH\":20,\"TEXT_MATCH\":21}}}},\"Query\":{\"fields\":{\"root\":{\"type\":\"Predicate\",\"id\":1}}},\"Value\":{\"oneofs\":{\"Type\":{\"oneof\":[\"null\",\"bool\",\"int\",\"float\",\"string\",\"timestamp\",\"datetime\",\"bytes\",\"object\",\"array\"]}},\"fields\":{\"null\":{\"type\":\"bool\",\"id\":1},\"bool\":{\"type\":\"bool\",\"id\":2},\"int\":{\"type\":\"int32\",\"id\":3},\"float\":{\"type\":\"float\",\"id\":4},\"string\":{\"type\":\"string\",\"id\":5},\"timestamp\":{\"type\":\"string\",\"id\":10},\"datetime\":{\"type\":\"string\",\"id\":11},\"bytes\":{\"type\":\"bytes\",\"id\":12},\"object\":{\"type\":\"Object\",\"id\":20},\"array\":{\"type\":\"Array\",\"id\":21}}}}}}}}}}}");
export const schema = Schema.fromJson<TYPES>(schemaJson);
