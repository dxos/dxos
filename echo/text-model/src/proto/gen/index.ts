import { Schema } from "@dxos/codec-protobuf";
import * as dxos_echo_text from "./dxos/echo/text";
export interface TYPES {
    "dxos.echo.text.Mutation": dxos_echo_text.Mutation;
    "dxos.echo.text.Snapshot": dxos_echo_text.Snapshot;
}
export const schemaJson = JSON.parse("{\"nested\":{\"dxos\":{\"nested\":{\"echo\":{\"nested\":{\"text\":{\"nested\":{\"Mutation\":{\"fields\":{\"update\":{\"type\":\"bytes\",\"id\":1},\"clientId\":{\"type\":\"int32\",\"id\":2}}},\"Snapshot\":{\"fields\":{\"data\":{\"type\":\"bytes\",\"id\":1}}}}}}}}}}}");
export const schema = Schema.fromJson<TYPES>(schemaJson);
