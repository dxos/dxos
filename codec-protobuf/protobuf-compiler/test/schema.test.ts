import { schema } from './gen';
import { ComplexFields, TaskList, TaskType } from './gen/dxos/test'
import { MyKey } from './my-key';
import { readFileSync, readdirSync, lstatSync } from 'fs'
import { join } from 'path'

test('encode and decode', async () => {
  const codec = schema.getCodecForType('dxos.test.TaskList')

  const initial: TaskList = {
    tasks: [
      {
        id: 'foo',
        title: 'Bar',
        key: new MyKey(Buffer.from('foo')),
        type: TaskType.COMPLETED,
        googleAny: {
          typeUrl: 'dxos.test.SubstitutedByInterface',
          foo: 'foo',
        }
      },
      {
        id: 'baz',
        title: 'Baz',
        key: new MyKey(Buffer.from('foo')),
        type: TaskType.IN_PROGRESS,
      }
    ],
  }

  const encoded = codec.encode(initial)
  
  expect(encoded).toBeInstanceOf(Uint8Array);

  const decoded = codec.decode(encoded)

  expect(decoded).toEqual(initial)
})

test('complex fields round trip', () => {
  const codec = schema.getCodecForType('dxos.test.ComplexFields');

  const initial: ComplexFields = {
    repeatedField: [new MyKey(Buffer.from('foo')), new MyKey(Buffer.from('bar'))],
    requiredField: new MyKey(Buffer.from('foo')),
    mappedField: {
      foo: new MyKey(Buffer.from('foo')),
      bar: new MyKey(Buffer.from('foo')),
    },
    inner: {
      bar: ComplexFields.InnerEnum.BAR,
      foo: 'foo'
    },
    myAny: {
      foo: 'foo'
    },
    googleAny: {
      typeUrl: 'dxos.test.Task',
      id: 'baz',
      title: 'Baz',
      key: new MyKey(Buffer.from('foo')),
      type: TaskType.IN_PROGRESS,
    },
    importedAny: {
      bar: 123,
    },
  }

  const encoded = codec.encode(initial)
  
  expect(encoded).toBeInstanceOf(Uint8Array);

  const decoded = codec.decode(encoded)

  expect(decoded).toEqual(initial)
})

test('definitions', () => {
  expect(readDirectoryFiles(join(__dirname, 'gen'))).toMatchSnapshot()
})


function readDirectoryFiles(dir: string) {
  let res = ''
  for(const file of listFilesRecursive(dir)) {
    res += `${file}:\n`
    res += readFileSync(join(dir, file), { encoding: 'utf-8' })
    res += '\n'
  }
  return res
}

function* listFilesRecursive(dir: string): Generator<string> {
  for (const file of readdirSync(dir)) {
    if(lstatSync(join(dir, file)).isDirectory()) {
      for(const sub of listFilesRecursive(join(dir, file))) {
        yield join(file, sub)
      }
    } else {
      yield file;
    }
  }
}
