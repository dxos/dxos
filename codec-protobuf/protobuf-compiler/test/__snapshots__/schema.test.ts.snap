// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`definitions 1`] = `
"dxos/test/another.ts:
import { MyKey } from \\"../../../my-key\\";
import { MyInterfaceSubstitution } from \\"../../../my-interface\\";
import { Schema as CodecSchema } from \\"@dxos/codec-protobuf\\";
import { DecodedAny } from \\"../../../any\\";
import * as dxos_test from \\"../test\\";
import * as dxos_test_any from \\"./any\\";
import * as google_protobuf from \\"../../google/protobuf\\";
export interface AnotherMessage {
}

dxos/test/any.ts:
import { MyKey } from \\"../../../my-key\\";
import { MyInterfaceSubstitution } from \\"../../../my-interface\\";
import { Schema as CodecSchema } from \\"@dxos/codec-protobuf\\";
import { DecodedAny } from \\"../../../any\\";
import * as dxos_test from \\"../test\\";
import * as dxos_test_another from \\"./another\\";
import * as google_protobuf from \\"../../google/protobuf\\";
export interface Any {
    bar?: number;
}

dxos/test.ts:
import { MyKey } from \\"../../my-key\\";
import { MyInterfaceSubstitution } from \\"../../my-interface\\";
import { Schema as CodecSchema } from \\"@dxos/codec-protobuf\\";
import { DecodedAny } from \\"../../any\\";
import * as dxos_test_another from \\"./test/another\\";
import * as dxos_test_any from \\"./test/any\\";
import * as google_protobuf from \\"../google/protobuf\\";
export interface Key {
    data: Uint8Array;
}
export interface Any {
    foo?: string;
}
export interface Scalars {
    doubleField?: number;
    floatField?: number;
    int32Field?: number;
    int64Field?: number;
    uint32Field?: number;
    uint64Field?: number;
    sint32Field?: number;
    sint64Field?: number;
    fixed32Field?: number;
    fixed64Field?: number;
    sfixed32Field?: number;
    sfixed64Field?: number;
    boolField?: boolean;
    stringField?: string;
    bytesField?: Uint8Array;
}
export interface ComplexFields {
    repeatedField?: MyKey[];
    requiredField: MyKey;
    mappedField?: Partial<Record<string, MyKey>>;
    inner?: ComplexFields.Inner;
    myAny?: Any;
    googleAny?: DecodedAny;
    importedAny?: dxos_test_any.Any;
}
export namespace ComplexFields {
    export interface Inner {
        foo?: string;
        bar?: ComplexFields.InnerEnum;
    }
    export enum InnerEnum {
        FOO = 1,
        BAR = 2
    }
}
export interface Outer {
    inner?: ComplexFields.Inner;
}
export namespace Outer {
    export interface Inner {
        inner?: ComplexFields.InnerEnum;
    }
}
export enum TaskType {
    IN_PROGRESS = 1,
    COMPLETED = 2
}
export interface Task {
    id?: string;
    title?: string;
    key: MyKey;
    type: TaskType;
    googleAny?: DecodedAny;
}
export interface TaskList {
    tasks?: Task[];
}
export interface SubstitutedByInterface {
    foo?: string;
}

google/protobuf.ts:
import { MyKey } from \\"../../my-key\\";
import { MyInterfaceSubstitution } from \\"../../my-interface\\";
import { Schema as CodecSchema } from \\"@dxos/codec-protobuf\\";
import { DecodedAny } from \\"../../any\\";
import * as dxos_test from \\"../dxos/test\\";
import * as dxos_test_another from \\"../dxos/test/another\\";
import * as dxos_test_any from \\"../dxos/test/any\\";
export interface Any {
    type_url?: string;
    value?: Uint8Array;
}

index.ts:
import { Schema } from \\"@dxos/codec-protobuf\\";
import substitutions from \\"../substitutions\\";
import { MyKey } from \\"../my-key\\";
import { MyInterfaceSubstitution } from \\"../my-interface\\";
import { Schema as CodecSchema } from \\"@dxos/codec-protobuf\\";
import { DecodedAny } from \\"../any\\";
import * as dxos_test from \\"./dxos/test\\";
import * as dxos_test_another from \\"./dxos/test/another\\";
import * as dxos_test_any from \\"./dxos/test/any\\";
import * as google_protobuf from \\"./google/protobuf\\";
export interface TYPES {
    \\"dxos.test.Key\\": dxos_test.Key;
    \\"dxos.test.Any\\": dxos_test.Any;
    \\"dxos.test.Scalars\\": dxos_test.Scalars;
    \\"dxos.test.ComplexFields\\": dxos_test.ComplexFields;
    \\"dxos.test.ComplexFields.Inner\\": dxos_test.ComplexFields.Inner;
    \\"dxos.test.ComplexFields.InnerEnum\\": dxos_test.ComplexFields.InnerEnum;
    \\"dxos.test.Outer\\": dxos_test.Outer;
    \\"dxos.test.Outer.Inner\\": dxos_test.Outer.Inner;
    \\"dxos.test.TaskType\\": dxos_test.TaskType;
    \\"dxos.test.Task\\": dxos_test.Task;
    \\"dxos.test.TaskList\\": dxos_test.TaskList;
    \\"dxos.test.SubstitutedByInterface\\": dxos_test.SubstitutedByInterface;
    \\"dxos.test.another.AnotherMessage\\": dxos_test_another.AnotherMessage;
    \\"dxos.test.any.Any\\": dxos_test_any.Any;
    \\"google.protobuf.Any\\": google_protobuf.Any;
}
export const schemaJson = JSON.parse(\\"{\\\\\\"nested\\\\\\":{\\\\\\"dxos\\\\\\":{\\\\\\"nested\\\\\\":{\\\\\\"test\\\\\\":{\\\\\\"nested\\\\\\":{\\\\\\"Key\\\\\\":{\\\\\\"fields\\\\\\":{\\\\\\"data\\\\\\":{\\\\\\"rule\\\\\\":\\\\\\"required\\\\\\",\\\\\\"type\\\\\\":\\\\\\"bytes\\\\\\",\\\\\\"id\\\\\\":1}}},\\\\\\"Any\\\\\\":{\\\\\\"fields\\\\\\":{\\\\\\"foo\\\\\\":{\\\\\\"type\\\\\\":\\\\\\"string\\\\\\",\\\\\\"id\\\\\\":1}}},\\\\\\"Scalars\\\\\\":{\\\\\\"fields\\\\\\":{\\\\\\"doubleField\\\\\\":{\\\\\\"type\\\\\\":\\\\\\"double\\\\\\",\\\\\\"id\\\\\\":1},\\\\\\"floatField\\\\\\":{\\\\\\"type\\\\\\":\\\\\\"float\\\\\\",\\\\\\"id\\\\\\":2},\\\\\\"int32Field\\\\\\":{\\\\\\"type\\\\\\":\\\\\\"int32\\\\\\",\\\\\\"id\\\\\\":3},\\\\\\"int64Field\\\\\\":{\\\\\\"type\\\\\\":\\\\\\"int64\\\\\\",\\\\\\"id\\\\\\":4},\\\\\\"uint32Field\\\\\\":{\\\\\\"type\\\\\\":\\\\\\"uint32\\\\\\",\\\\\\"id\\\\\\":5},\\\\\\"uint64Field\\\\\\":{\\\\\\"type\\\\\\":\\\\\\"uint64\\\\\\",\\\\\\"id\\\\\\":6},\\\\\\"sint32Field\\\\\\":{\\\\\\"type\\\\\\":\\\\\\"sint32\\\\\\",\\\\\\"id\\\\\\":7},\\\\\\"sint64Field\\\\\\":{\\\\\\"type\\\\\\":\\\\\\"sint64\\\\\\",\\\\\\"id\\\\\\":8},\\\\\\"fixed32Field\\\\\\":{\\\\\\"type\\\\\\":\\\\\\"fixed32\\\\\\",\\\\\\"id\\\\\\":9},\\\\\\"fixed64Field\\\\\\":{\\\\\\"type\\\\\\":\\\\\\"fixed64\\\\\\",\\\\\\"id\\\\\\":10},\\\\\\"sfixed32Field\\\\\\":{\\\\\\"type\\\\\\":\\\\\\"sfixed32\\\\\\",\\\\\\"id\\\\\\":11},\\\\\\"sfixed64Field\\\\\\":{\\\\\\"type\\\\\\":\\\\\\"sfixed64\\\\\\",\\\\\\"id\\\\\\":12},\\\\\\"boolField\\\\\\":{\\\\\\"type\\\\\\":\\\\\\"bool\\\\\\",\\\\\\"id\\\\\\":13},\\\\\\"stringField\\\\\\":{\\\\\\"type\\\\\\":\\\\\\"string\\\\\\",\\\\\\"id\\\\\\":14},\\\\\\"bytesField\\\\\\":{\\\\\\"type\\\\\\":\\\\\\"bytes\\\\\\",\\\\\\"id\\\\\\":15}}},\\\\\\"ComplexFields\\\\\\":{\\\\\\"fields\\\\\\":{\\\\\\"repeatedField\\\\\\":{\\\\\\"rule\\\\\\":\\\\\\"repeated\\\\\\",\\\\\\"type\\\\\\":\\\\\\"Key\\\\\\",\\\\\\"id\\\\\\":16},\\\\\\"requiredField\\\\\\":{\\\\\\"rule\\\\\\":\\\\\\"required\\\\\\",\\\\\\"type\\\\\\":\\\\\\"Key\\\\\\",\\\\\\"id\\\\\\":17},\\\\\\"mappedField\\\\\\":{\\\\\\"keyType\\\\\\":\\\\\\"string\\\\\\",\\\\\\"type\\\\\\":\\\\\\"Key\\\\\\",\\\\\\"id\\\\\\":18},\\\\\\"inner\\\\\\":{\\\\\\"type\\\\\\":\\\\\\"Inner\\\\\\",\\\\\\"id\\\\\\":19},\\\\\\"myAny\\\\\\":{\\\\\\"type\\\\\\":\\\\\\"Any\\\\\\",\\\\\\"id\\\\\\":20},\\\\\\"googleAny\\\\\\":{\\\\\\"type\\\\\\":\\\\\\"google.protobuf.Any\\\\\\",\\\\\\"id\\\\\\":21},\\\\\\"importedAny\\\\\\":{\\\\\\"type\\\\\\":\\\\\\"dxos.test.any.Any\\\\\\",\\\\\\"id\\\\\\":22}},\\\\\\"nested\\\\\\":{\\\\\\"Inner\\\\\\":{\\\\\\"fields\\\\\\":{\\\\\\"foo\\\\\\":{\\\\\\"type\\\\\\":\\\\\\"string\\\\\\",\\\\\\"id\\\\\\":1},\\\\\\"bar\\\\\\":{\\\\\\"type\\\\\\":\\\\\\"InnerEnum\\\\\\",\\\\\\"id\\\\\\":2}}},\\\\\\"InnerEnum\\\\\\":{\\\\\\"values\\\\\\":{\\\\\\"FOO\\\\\\":1,\\\\\\"BAR\\\\\\":2}}}},\\\\\\"Outer\\\\\\":{\\\\\\"fields\\\\\\":{\\\\\\"inner\\\\\\":{\\\\\\"type\\\\\\":\\\\\\"ComplexFields.Inner\\\\\\",\\\\\\"id\\\\\\":1}},\\\\\\"nested\\\\\\":{\\\\\\"Inner\\\\\\":{\\\\\\"fields\\\\\\":{\\\\\\"inner\\\\\\":{\\\\\\"type\\\\\\":\\\\\\"ComplexFields.InnerEnum\\\\\\",\\\\\\"id\\\\\\":1}}}}},\\\\\\"TaskType\\\\\\":{\\\\\\"values\\\\\\":{\\\\\\"IN_PROGRESS\\\\\\":1,\\\\\\"COMPLETED\\\\\\":2}},\\\\\\"Task\\\\\\":{\\\\\\"fields\\\\\\":{\\\\\\"id\\\\\\":{\\\\\\"type\\\\\\":\\\\\\"string\\\\\\",\\\\\\"id\\\\\\":1},\\\\\\"title\\\\\\":{\\\\\\"type\\\\\\":\\\\\\"string\\\\\\",\\\\\\"id\\\\\\":2},\\\\\\"key\\\\\\":{\\\\\\"rule\\\\\\":\\\\\\"required\\\\\\",\\\\\\"type\\\\\\":\\\\\\"Key\\\\\\",\\\\\\"id\\\\\\":3},\\\\\\"type\\\\\\":{\\\\\\"rule\\\\\\":\\\\\\"required\\\\\\",\\\\\\"type\\\\\\":\\\\\\"TaskType\\\\\\",\\\\\\"id\\\\\\":4},\\\\\\"googleAny\\\\\\":{\\\\\\"type\\\\\\":\\\\\\"google.protobuf.Any\\\\\\",\\\\\\"id\\\\\\":5}}},\\\\\\"TaskList\\\\\\":{\\\\\\"fields\\\\\\":{\\\\\\"tasks\\\\\\":{\\\\\\"rule\\\\\\":\\\\\\"repeated\\\\\\",\\\\\\"type\\\\\\":\\\\\\"Task\\\\\\",\\\\\\"id\\\\\\":1}}},\\\\\\"SubstitutedByInterface\\\\\\":{\\\\\\"fields\\\\\\":{\\\\\\"foo\\\\\\":{\\\\\\"type\\\\\\":\\\\\\"string\\\\\\",\\\\\\"id\\\\\\":1}}},\\\\\\"another\\\\\\":{\\\\\\"nested\\\\\\":{\\\\\\"AnotherMessage\\\\\\":{\\\\\\"fields\\\\\\":{}}}},\\\\\\"any\\\\\\":{\\\\\\"nested\\\\\\":{\\\\\\"Any\\\\\\":{\\\\\\"fields\\\\\\":{\\\\\\"bar\\\\\\":{\\\\\\"type\\\\\\":\\\\\\"int32\\\\\\",\\\\\\"id\\\\\\":1}}}}}}}}},\\\\\\"google\\\\\\":{\\\\\\"nested\\\\\\":{\\\\\\"protobuf\\\\\\":{\\\\\\"nested\\\\\\":{\\\\\\"Any\\\\\\":{\\\\\\"fields\\\\\\":{\\\\\\"type_url\\\\\\":{\\\\\\"type\\\\\\":\\\\\\"string\\\\\\",\\\\\\"id\\\\\\":1},\\\\\\"value\\\\\\":{\\\\\\"type\\\\\\":\\\\\\"bytes\\\\\\",\\\\\\"id\\\\\\":2}}}}}}}}}\\");
export const schema = Schema.fromJson<TYPES>(schemaJson, substitutions);

"
`;
